### ISSUE TITLE
DevContainers: Profile persistence

### ISSUE THREAD

**Issue by max06** [microsoft/vscode-remote-release#5823]
Something to share for your docs. It took me way too much time to figure out how to persist the user profile in a devcontainer in a generic way, might save some time for others...

```json
	"mounts": [
		"source=profile,target=/root,type=volume",
		"target=/root/.vscode-server,type=volume"
	],
```

Creates a named volume `profile` mounted to `/root`. Survives a rebuild.
Creates an anonymous volume mounted to `/root/.vscode-server`. Gets destroyed on rebuild, which allows vscode to reinstall the extensions and dotfiles.

**Comment by bamurtaugh** [comment]
Thanks for opening @max06! I think this may be addressed through https://code.visualstudio.com/remote/advancedcontainers/avoid-extension-reinstalls, which also describes creating a local "named volume" mount? 

If you feel like this request is different, or the current doc could have updated info/be more discoverable, would be happy to discuss further!

**Comment by max06** [comment]
Almost. The documentation describes how to keep only the installed extensions, 

My approach here is the exact opposite: I want to keep the user profile (`/root`) since we're using the devcontainers for an "operational" purpose. Means, we want to keep the shell history, tool configurations and similar stuff - while keeping the ability to update the devcontainer and the extensions. This scenario is not covered yet, to my knowledge.

**Comment by bamurtaugh** [comment]
Thanks for clarifying. So the documentation outlines how to persist extensions you've installed in your dev container and VS Code Server on rebuilds, and it won't delete the named volume. Your approach is intending to persist the user profile (shell history, tool config) through `/root` but update/reinstall extensions on rebuild with a fresh volume. 

Out of curiosity, what has been your main workflow where you've wanted to use this, i.e. are you able to expand on what it means to use a dev container for an "operational" purpose?

**Comment by max06** [comment]
Sure!

Me and my team are managing a lot of servers using ansible and the openstack-cli. The second one requires a configuration file with credentials (`/root/.config/openstack/clouds.yaml`). 

To make the installation process easier, we've added a devcontainer configuration containing everything required for this, like ansible and its dependencies, the openstack cli and modules, and some other tools we use on a daily basis. We also have different preferences when it comes to the shell environment. I'm using fish, others are working with bash or zsh - configured using personal dotfiles-repositories. Those shells are using different files for their histories, which makes using mounts more complicated/impossible. And not all of us are keeping their environments in sync with their dotfiles-repos. (And due to some other issues with the extension, I currently can't use mounts at all.)

The workflow is simple: We simply use the terminal in vscode, or connect to the container with a regular terminal application. Either way, we have a unified work environment, guaranteed to work for everyone. We even use the exact same image in our CI pipelines for automated parts of our stuff.

The main goal here is to achive updating/replacing the devcontainer with a newer version, in case we're adding a new tool to it for everybody to use - or a new vscode extension. VSCode asks for a rebuild, which usually results in a new, clean container. Without history, without tool configurations like the mentioned clouds.yaml. 

I should mention that the histories are basically a brain extension for us. We don't need to remember or lookup all the different switches and options of our tools, because we did it before, and can simply search for it. Persisting the complete user profile solves this, and even eliminates the need for custom shell configs/symlinks to store the data on a named volume. People are less reluctant to click the "rebuild now"-button since they are not afraid to loose their profiles anymore.
